### ISSUE TITLE
"properties" argument for "load_collection" in context of SHub Batch Process throws ClassCastException

### ISSUE THREAD

**Issue by bossie** [Open-EO/openeo-geopyspark-driver#131]
In an attempt to speed up the test_sentinel_hub_sar_backscatter_batch_process integration test, I noticed that support for the "properties" argument for "load_collection" is broken in the context of SHub Batch processes:

```[2022-03-15 07:50:58,768] INFO in openeogeotrellis.backend: Dry run extracted these source constraints: [(('load_collection', ('SENTINEL1_GAMMA0_SENTINELHUB', (('timeliness', (('eq', 'NRT3h'),)),))), {'temporal_extent': ('2019-10-10', '2019-10-10'), 'spatial_extent': {'west': 2.59003, 'south': 51.069, 'east': 2.8949, 'north': 51.2206, 'crs': 'EPSG:4326'}, 'bands': ['VV', 'VH'], 'sar_backscatter': SarBackscatterArgs(coefficient='gamma0-terrain', elevation_model='COPERNICUS_30', mask=True, contributing_area=False, local_incidence_angle=True, ellipsoid_incidence_angle=False, noise_removal=True, options={}), 'properties': {'timeliness': {'process_graph': {'eq1': {'process_id': 'eq', 'arguments': {'x': {'from_parameter': 'value'}, 'y': 'NRT3h'}, 'result': True}}}}})]
/data3/hadoop/yarn/local/usercache/openeo/appcache/application_1647270439955_0081/container_e5017_1647270439955_0081_01_000001/venv/lib/python3.8/site-packages/pyproj/crs/crs.py:280: FutureWarning: '+init=<authority>:<code>' syntax is deprecated. '<authority>:<code>' is the preferred initialization method. When making the change, be mindful of axis order changes: https://pyproj4.github.io/pyproj/stable/gotchas.html#axis-order-changes-in-proj-6
  projstring = _prepare_from_string(projparams)
/data3/hadoop/yarn/local/usercache/openeo/appcache/application_1647270439955_0081/container_e5017_1647270439955_0081_01_000001/venv/lib/python3.8/site-packages/pyproj/crs/crs.py:280: FutureWarning: '+init=<authority>:<code>' syntax is deprecated. '<authority>:<code>' is the preferred initialization method. When making the change, be mindful of axis order changes: https://pyproj4.github.io/pyproj/stable/gotchas.html#axis-order-changes-in-proj-6
  projstring = _prepare_from_string(projparams)
[2022-03-15 07:50:59,436] INFO in openeogeotrellis.backend: deemed collection SENTINEL1_GAMMA0_SENTINELHUB request CARD4L compliant (SarBackscatterArgs(coefficient='gamma0-terrain', elevation_model='COPERNICUS_30', mask=True, contributing_area=False, local_incidence_angle=True, ellipsoid_incidence_angle=False, noise_removal=True, options={}))
22/03/15 07:50:59 DEBUG geotrellissentinelhub.AuthApi: requesting new access token for client ID b667f624-3330-4bb8-a69e-6b0700d518a7
22/03/15 07:50:59 DEBUG geotrellissentinelhub.AuthApi: requesting new access token for client ID b667f624-3330-4bb8-a69e-6b0700d518a7
22/03/15 07:50:59 DEBUG geotrellissentinelhub.AuthApi: requesting new access token for client ID b667f624-3330-4bb8-a69e-6b0700d518a7
[2022-03-15 07:50:59,716] INFO in kazoo.client: Closing connection to epod-master3.vgt.vito.be:2181
[2022-03-15 07:50:59,716] INFO in kazoo.client: Zookeeper session closed, state: CLOSED
[2022-03-15 07:50:59,748] ERROR in openeo_driver.views: An error occurred while calling o1462.start_card4l_batch_processes.
: java.lang.ClassCastException: class java.util.ArrayList cannot be cast to class java.util.Map (java.util.ArrayList and java.util.Map are in module java.base of loader 'bootstrap')
	at org.openeo.geotrellissentinelhub.Criteria$.$anonfun$toQueryProperties$2(Criteria.scala:9)
	at scala.collection.TraversableLike$WithFilter.$anonfun$map$2(TraversableLike.scala:935)
	at scala.collection.Iterator.foreach(Iterator.scala:943)
	at scala.collection.Iterator.foreach$(Iterator.scala:943)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
	at scala.collection.TraversableLike$WithFilter.map(TraversableLike.scala:934)
	at org.openeo.geotrellissentinelhub.Criteria$.toQueryProperties(Criteria.scala:9)
	at org.openeo.geotrellissentinelhub.BatchProcessingService.$anonfun$start_card4l_batch_processes$1(BatchProcessingService.scala:166)
	at org.openeo.geotrellissentinelhub.package$.authorized(package.scala:63)
	at org.openeo.geotrellissentinelhub.BatchProcessingService.authorized(BatchProcessingService.scala:21)
	at org.openeo.geotrellissentinelhub.BatchProcessingService.start_card4l_batch_processes(BatchProcessingService.scala:164)
	at org.openeo.geotrellissentinelhub.BatchProcessingService.start_card4l_batch_processes(BatchProcessingService.scala:149)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)
	at py4j.ClientServerConnection.run(ClientServerConnection.java:106)
	at java.base/java.lang.Thread.run(Thread.java:834)
Traceback (most recent call last):
  File "/data3/hadoop/yarn/local/usercache/openeo/appcache/application_1647270439955_0081/container_e5017_1647270439955_0081_01_000001/venv/lib/python3.8/site-packages/flask/app.py", line 1516, in full_dispatch_request
    rv = self.dispatch_request()
  File "/data3/hadoop/yarn/local/usercache/openeo/appcache/application_1647270439955_0081/container_e5017_1647270439955_0081_01_000001/venv/lib/python3.8/site-packages/flask/app.py", line 1502, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
  File "/data3/hadoop/yarn/local/usercache/openeo/appcache/application_1647270439955_0081/container_e5017_1647270439955_0081_01_000001/venv/lib/python3.8/site-packages/openeo_driver/users/auth.py", line 84, in decorated
    return f(*args, **kwargs)
  File "/data3/hadoop/yarn/local/usercache/openeo/appcache/application_1647270439955_0081/container_e5017_1647270439955_0081_01_000001/venv/lib/python3.8/site-packages/openeo_driver/views.py", line 780, in queue_job
    backend_implementation.batch_jobs.start_job(job_id=job_id, user=user)
  File "/data3/hadoop/yarn/local/usercache/openeo/appcache/application_1647270439955_0081/container_e5017_1647270439955_0081_01_000001/venv/lib/python3.8/site-packages/openeogeotrellis/backend.py", line 882, in start_job
    self._start_job(job_id, user.user_id)
  File "/data3/hadoop/yarn/local/usercache/openeo/appcache/application_1647270439955_0081/container_e5017_1647270439955_0081_01_000001/venv/lib/python3.8/site-packages/openeogeotrellis/backend.py", line 910, in _start_job
    and self._scheduled_sentinelhub_batch_processes(spec['process_graph'], api_version, registry,
  File "/data3/hadoop/yarn/local/usercache/openeo/appcache/application_1647270439955_0081/container_e5017_1647270439955_0081_01_000001/venv/lib/python3.8/site-packages/openeogeotrellis/backend.py", line 1314, in _scheduled_sentinelhub_batch_processes
    batch_request_ids = list(batch_processing_service.start_card4l_batch_processes(
  File "/usr/local/spark/python/lib/py4j-0.10.9.2-src.zip/py4j/java_gateway.py", line 1309, in __call__
    return_value = get_return_value(
  File "/usr/local/spark/python/lib/py4j-0.10.9.2-src.zip/py4j/protocol.py", line 326, in get_return_value
    raise Py4JJavaError(
py4j.protocol.Py4JJavaError: An error occurred while calling o1462.start_card4l_batch_processes.
: java.lang.ClassCastException: class java.util.ArrayList cannot be cast to class java.util.Map (java.util.ArrayList and java.util.Map are in module java.base of loader 'bootstrap')
	at org.openeo.geotrellissentinelhub.Criteria$.$anonfun$toQueryProperties$2(Criteria.scala:9)
	at scala.collection.TraversableLike$WithFilter.$anonfun$map$2(TraversableLike.scala:935)
	at scala.collection.Iterator.foreach(Iterator.scala:943)
	at scala.collection.Iterator.foreach$(Iterator.scala:943)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
	at scala.collection.TraversableLike$WithFilter.map(TraversableLike.scala:934)
	at org.openeo.geotrellissentinelhub.Criteria$.toQueryProperties(Criteria.scala:9)
	at org.openeo.geotrellissentinelhub.BatchProcessingService.$anonfun$start_card4l_batch_processes$1(BatchProcessingService.scala:166)
	at org.openeo.geotrellissentinelhub.package$.authorized(package.scala:63)
	at org.openeo.geotrellissentinelhub.BatchProcessingService.authorized(BatchProcessingService.scala:21)
	at org.openeo.geotrellissentinelhub.BatchProcessingService.start_card4l_batch_processes(BatchProcessingService.scala:164)
	at org.openeo.geotrellissentinelhub.BatchProcessingService.start_card4l_batch_processes(BatchProcessingService.scala:149)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:182)
	at py4j.ClientServerConnection.run(ClientServerConnection.java:106)
	at java.base/java.lang.Thread.run(Thread.java:834)
```
Request:

```python
    sar_backscatter = (auth_connection
                       .load_collection('SENTINEL1_GAMMA0_SENTINELHUB', bands=["VV", "VH"],
                                        properties={"timeliness": lambda t: eq(t, "NRT3h")})
                       .filter_bbox(west=2.59003, east=2.8949, north=51.2206, south=51.069)
                       .filter_temporal(extent=["2019-10-10", "2019-10-10"])
                       .sar_backscatter(mask=True, local_incidence_angle=True, elevation_model='COPERNICUS_30'))

    job = sar_backscatter.execute_batch(out_format='GTiff', title="SentinelhubSarBackscatterBatch")
```

The signature of org.openeo.geotrellissentinelhub.BatchProcessingService#start_card4l_batch_processes has changed to support more operators in addition to "eq" in https://jira.vito.be/browse/EP-4102.

**Comment by bossie** [comment]
Branch  https://github.com/Open-EO/openeo-geopyspark-driver/tree/%23131_properties_SHub_batch_process fixes the signature change but needs additional work (logs batch_job.py-:

```
2022-03-15 08:49:17,368:P81:INFO:openeogeotrellis.deploy.batch_job:batch_job.py main: start 2022-03-15 08:49:17.368398
2022-03-15 08:49:17,368:P81:INFO:openeogeotrellis.deploy.batch_job:batch_job.py main: start 2022-03-15 08:49:17.368398
2022-03-15 08:49:17,368:P81:INFO:openeogeotrellis.deploy.batch_job:argv: ['batch_job.py', 'f8822389-e1cd-4a1c-acf8-a618df86cbbc___3os_o7.in', '/data/projects/OpenEO/f8822389-e1cd-4a1c-acf8-a618df86cbbc', 'out', 'log', 'job_metadata.json', '1.0.0', '[{"collection_id": "SENTINEL1_GAMMA0_SENTINELHUB", "metadata_properties": {"timeliness": {"eq": "NRT3h", "source_location": "s3://openeo-sentinelhub/b613ec21-b964-41d9-87e9-673bd4f819de", "card4l": true}]', 'vdboschj']
2022-03-15 08:49:17,368:P81:INFO:openeogeotrellis.deploy.batch_job:argv: ['batch_job.py', 'f8822389-e1cd-4a1c-acf8-a618df86cbbc___3os_o7.in', '/data/projects/OpenEO/f8822389-e1cd-4a1c-acf8-a618df86cbbc', 'out', 'log', 'job_metadata.json', '1.0.0', '[{"collection_id": "SENTINEL1_GAMMA0_SENTINELHUB", "metadata_properties": {"timeliness": {"eq": "NRT3h", "source_location": "s3://openeo-sentinelhub/b613ec21-b964-41d9-87e9-673bd4f819de", "card4l": true}]', 'vdboschj']
2022-03-15 08:49:17,368:P81:INFO:openeogeotrellis.deploy.batch_job:pid 81; ppid 24; cwd /data2/hadoop/yarn/local/usercache/vdboschj/appcache/application_1647270439955_1616/container_e5017_1647270439955_1616_01_000001
2022-03-15 08:49:17,368:P81:INFO:openeogeotrellis.deploy.batch_job:pid 81; ppid 24; cwd /data2/hadoop/yarn/local/usercache/vdboschj/appcache/application_1647270439955_1616/container_e5017_1647270439955_1616_01_000001
2022-03-15 08:49:17,368:P81:INFO:openeogeotrellis.deploy.batch_job:batch_job.py main: fail 2022-03-15 08:49:17.368808, elapsed 0:00:00.000410
2022-03-15 08:49:17,368:P81:INFO:openeogeotrellis.deploy.batch_job:batch_job.py main: fail 2022-03-15 08:49:17.368808, elapsed 0:00:00.000410
Traceback (most recent call last):
  File "batch_job.py", line 474, in <module>
    main(sys.argv)
  File "batch_job.py", line 241, in main
    dependencies = _deserialize_dependencies(argv[7])
  File "batch_job.py", line 216, in _deserialize_dependencies
    dependencies = json.loads(arg)
  File "/usr/lib64/python3.8/json/__init__.py", line 357, in loads
    return _default_decoder.decode(s)
  File "/usr/lib64/python3.8/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib64/python3.8/json/decoder.py", line 353, in raw_decode
    obj, end = self.scan_once(s, idx)
json.decoder.JSONDecodeError: Expecting ',' delimiter: line 1 column 204 (char 203)
```

**Comment by bossie** [comment]
This "dependencies" argument (= mapping load_collection to corresponding SHub batch process) is currently faulty so this ticket is blocked until https://github.com/Open-EO/openeo-geopyspark-driver/issues/150 is resolved.

**Comment by bossie** [comment]
Fix in #150 made this fix possible.
